### PR TITLE
Add OpenAI-compatible HTTP API

### PR DESCRIPTION
**Pull request by ZacharyZcR**
## Summary

- add `coli serve`, a dependency-free OpenAI-compatible HTTP gateway backed by one persistent `glm` process
- add a length-prefixed raw prompt command to the existing C serve protocol so complete multi-message prompts and embedded newlines are preserved
- reuse the longest matching token/KV prefix across stateless HTTP requests, prefill only the appended or divergent tail, and rewind the persisted `.coli_kv` header at the same boundary
- implement `GET /v1/models`, model retrieval, `POST /v1/chat/completions`, and legacy `POST /v1/completions`
- support normal JSON responses, SSE streaming, streaming usage, Bearer auth, `max_tokens` / `max_completion_tokens`, `temperature`, `top_p`, and GLM reasoning mode
- serialize generations through the single loaded 744B engine instead of loading duplicate model processes
- support browser/Tauri clients with OPTIONS preflight and an explicit CORS origin allowlist

The request and response shapes follow the official [Chat Completions](https://platform.openai.com/docs/api-reference/chat/completions/create) and [Models](https://platform.openai.com/docs/api-reference/models) APIs. Chat rendering follows the official GLM-5.2 `chat_template.jinja` text path.

## KV reuse

OpenAI-style clients resend the complete conversation on every request. The engine tokenizes that prompt into a scratch buffer, compares it with `hist`, keeps the longest common prefix in memory, and forwards only the new tail. If history was edited, both the in-memory length and persisted KV record count rewind to the first divergent token; the following append overwrites that tail. An identical prompt re-forwards its last token to reconstruct logits without a full prefill.

`STAT` keeps its original fields first and appends full prompt-token and length-limit fields. Both the interactive CLI parser and the HTTP gateway accept the extended format.

## Scope

This first version is deliberately text-only. Tools, multimodal content, custom stop sequences, log probabilities, token penalties, structured output, and per-request seeds return explicit errors. Concurrent HTTP requests queue on the persistent engine.

The API binds to localhost by default. `COLI_API_KEY` enables Bearer authentication. Vite and Tauri local origins are allowed by default; additional exact origins require repeated `--cors-origin` flags.

## Local validation

- `make check` — portable C build and 13 dependency-free tests pass
- official `openai` Python SDK round-trip against the gateway: models list, non-streaming chat, and streaming chat pass
- browser OPTIONS/CORS preflight test passes
- `python3 -m py_compile c/openai_server.py c/tests/test_openai_server.py c/coli`
- `git diff --check`

A real 744B model generation was not run in this workspace; the C engine protocol compiles cleanly, while the HTTP and streaming path is covered with a deterministic fake engine.

### CODE REVIEW

**Review by JustVugg — CHANGES_REQUESTED** [review]
This is exactly the shape #14 asked for — stdlib-only gateway, honest scope, careful protocol framing (bounds checks, temp/nuc restore, the `memchr(raw,0,..)` guard). One design issue needs fixing before merge, because it decides whether the API is actually usable for chat on the hardware this project targets:

**`\x02PROMPT` resets the KV on every request (`len=0; first=1`).** HTTP chat is stateless — each request re-sends the whole conversation — so every message triggers a full re-prefill from token 0. On the machines this engine runs on, prefill is the expensive part: a 1,000-token conversation would take minutes *per message*, growing linearly. The interactive serve path keeps KV persistent across turns precisely to avoid this.

**Fix: longest-common-prefix reuse instead of reset.** OpenAI-style requests grow by appending, and `hist` already contains the rendered conversation *including* the assistant replies the engine itself generated — so the new request's token stream is (near-)identical to `hist` up to the new user message. Sketch:

```c
/* raw_mode: tokenize into a scratch buffer, keep the common prefix in KV */
int *tmp=malloc(maxctx*sizeof(int));
int k=tok_encode(&T,input,input_n,tmp,maxctx);
int p=0; while(p<len && p<k && hist[p]==tmp[p]) p++;
if(p<len){ len=p; if(m->has_mtp) m->kv_start[m->c.n_layers]=-1; }  /* divergence: drop tail only */
memcpy(hist+len, tmp+len, (k-len)*sizeof(int));
k-=len;              /* prefill only the new tail; k==0 edge: re-forward last token like MORE */
free(tmp);
```

Worst case (client edited history, different template) the prefix is short and it degrades gracefully to today's behavior; typical case it prefills only the last user message. It also composes with the KV-to-disk persistence landing on main (conversations resume warm across engine restarts — your gateway would inherit that for free).

Second, smaller point: the `STAT` line grew two fields — the interactive `coli` parser needs to stay compatible (your coli patch handles it, just flagging the cross-PR coupling for anyone rebasing).

Everything else reads well. Fix the prefix reuse and this merges — it's the most-requested feature in the tracker.

**Comment by ZacharyZcR** [comment]
Updated per review and rebased onto the latest `main` (including `.coli_kv` persistence):

- raw prompts are tokenized into a scratch buffer and matched against `hist`
- the longest common KV prefix is retained; only the new tail is prefetched
- edited/shorter histories rewind both the in-memory context and persisted KV record count to the divergence point, so the next append overwrites only the tail
- identical prompts re-forward the last token to reconstruct logits (`k == 0`)
- extended `STAT` output keeps the original fields first and reports the full prompt-token count
- the interactive `coli` parser remains compatible with the extended status line

Validation after rebase: `make check` (portable C build + 13 tests), Python byte-compilation, `git diff --check`, and an official OpenAI Python SDK round-trip for models, non-streaming chat, and streaming chat. A real 744B inference run is still not available in this workspace.

**Review by JustVugg — APPROVED** [review]
Prefix-reuse verified end-to-end here (tiny model, real server + real engine): first request `[API] KV prefix 0/62, prefill 62`, second request extending the conversation `[API] KV prefix 62/120, prefill 58` — the previous conversation's KV survives stateless HTTP exactly as designed, and the `kv_disk_truncate` integration keeps `.coli_kv` coherent on divergence. /v1/models, non-streaming, SSE streaming and the k=0 identical-prompt path all behave. Engine untouched elsewhere: oracle 32/32 TF, 20/20 greedy. Great turnaround — merging.
